src/neighborsearch/bounded.jl

```diff
@@ -3,39 +3,122 @@
 # ------------------------------------------------------------------
 
 """
-    BoundedSearch(method, nmax)
+    BoundedSearch(method, maxneighbors=Inf; maxpercategory,
+    maxperoctant, ordermetric=Euclidean())
 
-A method for searching at most `nmax` neighbors using `method`.
+A method for searching at most `maxneighbors` neighbors using `method`. Extra
+restrictions available: `maxpercategory` and `maxperoctant`. The priority is
+given to the nearest neighbor using `ordermetric`.
 """
 struct BoundedSearch{M<:NeighborSearchMethod} <: BoundedNeighborSearchMethod
   method::M
-  nmax::Int
+  maxneighbors::Int
+  maxpercategory::Dict{Symbol,Int}
+  maxperoctant::Int
+  ordermetric::Metric
+
+  function BoundedSearch{M}(method::M, maxneighbors=0; maxpercategory=Dict(),
+                            maxperoctant=0, ordermetric=Euclidean()) where {M}
+    new(method, maxneighbors, maxpercategory, maxperoctant, ordermetric)
+  end
 end
 
+BoundedSearch(method::M, maxneighbors::Int=0; kwargs...) where {M} =
+  BoundedSearch{M}(method, maxneighbors; kwargs...)
+
 object(method::BoundedSearch) = object(method.method)
 
-maxneighbors(method::BoundedSearch) = method.nmax
+maxneighbors(method::BoundedSearch) = method.maxneighbors
 
 function search!(neighbors, xₒ::AbstractVector,
                  method::BoundedSearch; mask=nothing)
-  inds = search(xₒ, method.method)
-  nmax = method.nmax
+  # get reference objects and initial neighbors
+  meth = method.method
+  obj  = object(method)
+  N    = ncoords(obj)
+  inds = search(xₒ, meth, mask=mask)
+
+  # get distances and give priority to closest neighbors according to ordermetric
+  ometric = method.ordermetric
+  dists   = colwise(ometric, xₒ, coordinates(obj,inds))
+  sortids = sortperm(dists)
+  inds    = inds[sortids]
+
+  # read constraints
+  maxk   = method.maxneighbors
+  maxoct = method.maxperoctant
+  maxcat = method.maxpercategory
```

Comment:
I would move all these lines where we are simply accessing fields of the method to the top of the implementation. After all fields are read, we can then start the logic in a separate section, step by step. Appreciate if you can refactor this.